soliditygen/generator.go

```
@@ -0,0 +1,159 @@
+}
+
+// convertToSolidityStruct recursively generates a child-first list of Solidity struct definitions.
+func convertToSolidityStruct(structName string, value interface{}) []Definition {
+	obj, ok := value.(map[string]interface{})
+	if !ok {
+		fallback := fmt.Sprintf("struct %s {\n    string fallback;\n}", structName)
+		return []Definition{{
+			Name: structName,
+			Text: fallback,
+		}}
+	}
+
+	var childDefs []Definition
+	var lines []string
+
+	for key, val := range obj {
+		solType, nestedDefs := inferSolidityType(key, val)
+
+		childDefs = append(childDefs, nestedDefs...)
+		lines = append(lines, fmt.Sprintf("    %s %s;", solType, key))
+	}
+
+	structText := fmt.Sprintf("struct %s {\n%s\n}", structName, strings.Join(lines, "\n"))
+
+	return append(childDefs, Definition{
+		Name: structName,
+		Text: structText,
+	})
+}
+
+// inferSolidityType decides which Solidity type to use for a given JSON value,
+// optionally returning new nested struct definitions if the value is a nested object/array of objects.
+func inferSolidityType(fieldName string, v interface{}) (string, []Definition) {
+	switch val := v.(type) {
+	case string:
+		return "string", nil
+	case bool:
+		return "bool", nil
+	case float64:
+		if val == float64(int64(val)) {
+			if val < 0 {
+				return "int256", nil
+			}
+			return "uint256", nil
+		}
+		return "int256", nil
+	case map[string]interface{}:
+		nestedName := capitalize(fieldName)
+		childDefs := convertToSolidityStruct(nestedName, val)
+		return nestedName, childDefs
+	case []interface{}:
+		var firstNonNull interface{}
+		for _, elem := range val {
+			if elem != nil {
+				firstNonNull = elem
+				break
+			}
+		}
+		if firstNonNull == nil {
+			return "string[]", nil
+		}
+		elemType, nestedDefs := inferSolidityType(fieldName, firstNonNull)
+		return elemType + "[]", nestedDefs
+	case nil:
+		return "string", nil
+	default:
+		return "string", nil
+	}
+}
+
+func capitalize(str string) string {
+	if str == "" {
+		return "Empty"
```

i'd rather return an error or panic, otherwise you can't really tell if there's a bug somewhere in your code or if you're actually capitalizing the string "empty"